src/server/cmd/pachd/main.go

```diff
@@ -410,17 +426,102 @@ func doFullMode(appEnvObj interface{}) error {
 		return err
 	})
 	eg.Go(func() error {
-		http.Handle("/metrics", promhttp.Handler())
-		err := http.ListenAndServe(fmt.Sprintf(":%v", assets.PrometheusPort), nil)
+		err := grpcutil.Serve(
+			grpcutil.ServerSpec{
+				Port:       appEnv.PeerPort,
+				MaxMsgSize: grpcutil.MaxMsgSize,
+				RegisterFunc: func(s *grpc.Server) error {
+					cacheServer := cache_server.NewCacheServer(router, appEnv.NumShards)
+					go func() {
+						if err := sharder.RegisterFrontends(nil, address, []shard.Frontend{cacheServer}); err != nil {
+							log.Printf("error from sharder.RegisterFrontend %s", grpcutil.ScrubGRPC(err))
+						}
+					}()
+					go func() {
+						if err := sharder.Register(nil, address, []shard.Server{cacheServer}); err != nil {
+							log.Printf("error from sharder.Register %s", grpcutil.ScrubGRPC(err))
+						}
+					}()
+					cache_pb.RegisterGroupCacheServer(s, cacheServer)
+
+					blockCacheBytes, err := units.RAMInBytes(appEnv.BlockCacheBytes)
+					if err != nil {
+						return err
+					}
+					blockAPIServer, err := pfs_server.NewBlockAPIServer(
+						appEnv.StorageRoot, blockCacheBytes, appEnv.StorageBackend, etcdAddress)
+					if err != nil {
+						return err
+					}
+					pfsclient.RegisterObjectAPIServer(s, blockAPIServer)
+
+					// Unfortunately, calling Register___Server(x) twice on the same
+					// struct x doesn't work--x will only serve requests from the first
+					// grpc.Server it was registered with. So we create a second set of
+					// APIServer structs here so we can serve the Pachyderm API on the
+					// peer port
```

Comment:
This comment seems a bit out of place since it seems like within this block there's only one copy of each server. Am I missing something?

Reply:
Oh, I guess this this comment just refers to the block as a whole, maybe it should be above the `eg.Go` call?

Reply:
Done